File: social-service/src/main/java/com/app/socialservice/follow/application/repository/FollowRepository.java
package com.app.socialservice.follow.application.repository;

import java.util.Optional;
import java.util.UUID;

import com.app.socialservice.follow.domain.model.Follow;

public interface FollowRepository {
    Optional<Follow> findActiveByUsers(UUID followerUserId, UUID followedUserId);
    Optional<Follow> findRemovedByUsers(UUID followerUserId, UUID followedUserId);
    boolean existsBlockedByUsers(UUID followerUserId, UUID followedUserId);
    boolean insertIfAbsent(Follow follow);
    boolean reactivate(UUID followerUserId, UUID followedUserId);
    boolean markBidirectionalRelationshipsAsBlocked(UUID firstUserId, UUID secondUserId);
}

File: recommendation-service/src/follow/repository/follow_repository.py

from abc import ABC, abstractmethod
from uuid import UUID

from follow.model.follow import Follow


class FollowRepository(ABC):
    @abstractmethod
    def create_follow(self, follow: Follow) -> None:
        pass

    @abstractmethod
    def remove_follow(self, follower_user_id: UUID, followed_user_id: UUID) -> None:
        pass

    @abstractmethod
    def remove_follows_between_users(self, user_id_1: UUID, user_id_2: UUID) -> None:
        pass
